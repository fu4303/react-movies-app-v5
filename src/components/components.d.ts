interface GlobalStyleProps {
  fontColor: string;
  bodyColor: string;
}

interface NavbarProps {
  title: string;
}

interface MainMovieContainerProps {
  backgroundImage: string;
}

interface MovieCardProps {
  id: number;
  poster_path?: string;
  backdrop_path?: string;
  title: string;
  release_date: string;
  vote_average?: number;
  overview: string;
  isFavorite?: boolean;
  isOnWatchList?: boolean;
  fetchFavorites?: () => void;
  fetchWatchList?: () => void;
}

interface ButtonProps {
  id?: string;
  text?: string;
  bgColor?: string;
  fontColor?: string;
  icon?: string;
  responsive?: boolean;
  onClick?: () => void;
}

interface ButtonStyles {
  bgColor?: string;
  fontColor?: string;
  isResponsive?: boolean;
  onlyIcon?: boolean;
}

interface LinkButtonProps {
  bgColor?: string;
  fontColor?: string;
}

interface CategoryCardprops {
  title: string;
  bgColor: string;
  fontColor: string;
}

interface CategoryCardStyles {
  bgColor: string;
  fontColor: string;
}

interface NoDataProps {
  title: string;
  icon: string;
}

interface SearchInputProps {
  searchHandler: (query: string) => Promise<void> | void;
  placeholder?: string;
}

interface GridFrameProps {
  center?: boolean;
}

interface SweetAlertProps {
  title: string;
  text: string;
  onButtonText?: string;
  onOkAction?: (() => void) | undefined;
}

interface SweetAlertToastProps {
  title: string;
  type?: 'success' | 'error' | 'warning' | 'info' | 'question'
}
