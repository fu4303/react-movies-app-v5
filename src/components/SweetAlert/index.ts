import Swal from 'sweetalert2';
import { theme } from 'themes';

const Toast = Swal.mixin({
  toast: true,
  position: 'top-end',
  showConfirmButton: false,
  timer: 3000,
  timerProgressBar: true,
  onOpen: (toast) => {
    toast.addEventListener('mouseenter', Swal.stopTimer);
    toast.addEventListener('mouseleave', Swal.resumeTimer);
  },
});

export const ErrorAlert = ({ title, text }: SweetAlertProps) => {
  Swal.fire({
    icon: 'error',
    title,
    text,
    confirmButtonColor: theme.darkShallow,
  });
};

export const SuccessToast = ({ title, type = 'success' }: SweetAlertToastProps) => {
  Toast.fire({
    icon: type,
    title,
  });
};
