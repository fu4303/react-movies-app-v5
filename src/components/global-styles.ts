import styled, { createGlobalStyle } from 'styled-components';
import { ButtonBaseLayout } from 'components/Button/styles';

export const GlobalStyle = createGlobalStyle<GlobalStyleProps>`
  /* CSS Reset */
  html, body, div, span, applet, object, iframe,
  h1, h2, h3, h4, h5, h6, p, blockquote, pre,
  a, abbr, acronym, address, big, cite, code,
  del, dfn, em, img, ins, kbd, q, s, samp,
  small, strike, strong, sub, sup, tt, var,
  b, u, i, center,
  dl, dt, dd, ol, ul, li,
  fieldset, form, label, legend,
  table, caption, tbody, tfoot, thead, tr, th, td,
  article, aside, canvas, details, embed, 
  figure, figcaption, footer, header, hgroup, 
  menu, nav, output, ruby, section, summary,
  time, mark, audio, video {
    margin: 0;
    padding: 0;
    border: 0;
    font-size: 100%;
    font: inherit;
    vertical-align: baseline;
  }
  /* HTML5 display-role reset for older browsers */
  article, aside, details, figcaption, figure, 
  footer, header, hgroup, menu, nav, section {
    display: block;
  }
  body {
    line-height: 1;
    font-family: 'Open Sans', sans-serif;
    background-color: ${({ bodyColor }) => bodyColor};
    color: ${({ fontColor }) => fontColor};

    padding: 5rem 1rem 1rem 1rem;


    @media only screen and (max-width: 625px) {
      padding-top: 8rem;
    }
  }
  ol, ul {
    list-style: none;
  }
  blockquote, q {
    quotes: none;
  }
  blockquote:before, blockquote:after,
  q:before, q:after {
    content: '';
    content: none;
  }
  table {
    border-collapse: collapse;
    border-spacing: 0;
  }
`;

export const GridFrame = styled.section<GridFrameProps>`
  display: grid;
  grid-template-columns: repeat(auto-fill, minmax(300px, 1fr));
  grid-gap: 1.5rem;
  
  @media only screen and (max-width: 625px) {
    justify-items: center;
  }
`;

export const LinkButton = styled.a<LinkButtonProps>`
  ${ButtonBaseLayout}
  text-decoration: none;
  padding: 0.7rem 1rem;
  background-color: ${({ bgColor }) => bgColor};
  color: ${({ fontColor }) => fontColor};

  @media only screen and (max-width: 800px) {
    padding: 0;
  }
`;
