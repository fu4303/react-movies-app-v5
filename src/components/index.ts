export * from './Navbar/Navbar';
export * from './MainMovie/MainMovie';
export * from './Button/Button';
export * from './GridWithTitle/GridWithTitle';
export * from './MovieCard/MovieCard';
export * from './Footer/Footer';
export * from './SearchInput/SearchInput';
export * from './NoData/NoData';
export * from './SweetAlert/index';
