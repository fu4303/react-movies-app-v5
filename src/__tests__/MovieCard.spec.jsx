import React from 'react';
import { createMemoryHistory } from 'history';
import { MemoryRouter, Router } from 'react-router-dom';
import { render, cleanup, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import { MovieCard } from 'components';

const mockMovieProps = (favorite = false, onwatchlist = false) => ({
  poster_path: '/e1mjopzAS2KNsvpbpahQ1a6SkSn.jpg',
  overview: 'From DC Comics comes the Suicide Squad, an antihero team of incarcerated supervillains who act as deniable assets for the United States government, undertaking high-risk black ops missions in exchange for commuted prison sentences.',
  release_date: '2016-08-03',
  id: 297761,
  title: 'Suicide Squad',
  isFavorite: favorite,
  isOnWatchList: onwatchlist,
});

afterEach(cleanup);

describe('Movie Card component tests', () => {
  it('Should match the snapshot for "MovieCard" component', () => {
    const { asFragment } = render(
      <MemoryRouter>
        <MovieCard {...mockMovieProps()} />
      </MemoryRouter>,
    );
    expect(asFragment(<MovieCard {...mockMovieProps()} />)).toMatchSnapshot();
  });

  it('Should render component "MovieCard" correctly with props', () => {
    const { getByText, getByTestId } = render(
      <MemoryRouter>
        <MovieCard {...mockMovieProps()} />
      </MemoryRouter>,
    );

    expect(getByTestId('movieCardPoster')).toHaveAttribute('src', 'undefinedoriginal/e1mjopzAS2KNsvpbpahQ1a6SkSn.jpg');
    expect(getByText(/From DC Comics comes the Suicide Squad, an antihero team of incarcerated/i)).toBeInTheDocument();
    expect(getByText('2016')).toBeInTheDocument();
    expect(getByTestId('movieCardClicableTitle')).toHaveTextContent('Suicide Squad');
    expect(getByText('launch')).toBeInTheDocument();
    expect(getByText('playlist_add')).toBeInTheDocument();
    expect(getByText('star_border')).toBeInTheDocument();
  });

  it('Should navigate to route when click on image on "MovieCard"', () => {
    const history = createMemoryHistory();

    history.push = jest.fn();

    const { getByTestId } = render(
      <Router history={history}>
        <MovieCard {...mockMovieProps()} />
      </Router>,
    );

    fireEvent.click(getByTestId('movieCardClicableImage'));
    expect(history.push).toHaveBeenCalledWith('/movie/297761');
  });

  it('Should navigate to route when click on movie title on "MovieCard"', () => {
    const history = createMemoryHistory();

    history.push = jest.fn();

    const { getByTestId } = render(
      <Router history={history}>
        <MovieCard {...mockMovieProps()} />
      </Router>,
    );

    fireEvent.click(getByTestId('movieCardClicableTitle'));
    expect(history.push).toHaveBeenCalledWith('/movie/297761');
  });

  it('Should navigate to route when click on See Details button on "MovieCard"', () => {
    const history = createMemoryHistory();

    history.push = jest.fn();

    const { getByTestId } = render(
      <Router history={history}>
        <MovieCard {...mockMovieProps()} />
      </Router>,
    );

    fireEvent.click(getByTestId('movieCardGoToDetailsButton'));
    expect(history.push).toHaveBeenCalledWith('/movie/297761');
  });

  it('Should toggle icon on click on Add/Remove to Watch List button on "MovieCard"', () => {
    const { getByTestId } = render(
      <MemoryRouter>
        <MovieCard {...mockMovieProps()} />
      </MemoryRouter>,
    );

    fireEvent.click(getByTestId('movieCardToggleAddToList'));
    expect(getByTestId('movieCardToggleAddToList')).toHaveTextContent('playlist_add_check');
  });

  it('Should toggle icon on click on Add/Remove to favorites button on "MovieCard"', () => {
    const { getByTestId } = render(
      <MemoryRouter>
        <MovieCard {...mockMovieProps()} />
      </MemoryRouter>,
    );

    fireEvent.click(getByTestId('movieCardToggleAddToFavorites'));
    expect(getByTestId('movieCardToggleAddToFavorites')).toHaveTextContent('star');
  });
});
