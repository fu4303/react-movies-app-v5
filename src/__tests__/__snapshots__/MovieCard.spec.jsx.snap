// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Movie Card component tests Should match the snapshot for "MovieCard" component 1`] = `
<DocumentFragment>
  <div
    class="sc-fzpjYC iCmXBs"
  >
    <a
      data-testid="movieCardClicableImage"
      href="/movie/297761"
    >
      <img
        class="sc-fznJRM kzgVss"
        data-testid="movieCardPoster"
        src="undefinedoriginal/e1mjopzAS2KNsvpbpahQ1a6SkSn.jpg"
      />
    </a>
    <div
      class="sc-fznxsB cBLacw"
    >
      <a
        class="sc-fznWqX ccSiWP"
        data-testid="movieCardClicableTitle"
        href="/movie/297761"
      >
        Suicide Squad
      </a>
      <p
        class="sc-fzoiQi dDLToi"
      >
        2016
      </p>
      <p
        class="sc-fzqARJ hdNEPE"
      >
        From DC Comics comes the Suicide Squad, an antihero team of incarcerated supervillains who act as deniable assets for the...
      </p>
      <section
        class="sc-fzqNqU eSyPdJ"
      >
        <button
          class="sc-fzqNJr craSqe"
          data-testid="movieCardGoToDetailsButton"
          type="button"
        >
          <i
            class="material-icons"
          >
            launch
          </i>
        </button>
        <button
          class="sc-fzqNJr efUzvm"
          data-testid="movieCardToggleAddToList"
          type="button"
        >
          <i
            class="material-icons"
          >
            playlist_add
          </i>
        </button>
        <button
          class="sc-fzqNJr JyITD"
          data-testid="movieCardToggleAddToFavorites"
          type="button"
        >
          <i
            class="material-icons"
          >
            star_border
          </i>
        </button>
      </section>
    </div>
  </div>
</DocumentFragment>
`;
